apps/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -0,0 +1,588 @@
+                  'the workspace. Run "rush update" to update workspace references for this package.'
+              )
+            );
+            throw new AlreadyReportedError();
+          }
+
+          // We will update to `workspace` notation. If the version specified is a range, then use the provided range.
+          // Otherwise, use `workspace:*` to ensure we're always using the workspace package.
+          const workspaceRange: string =
+            !!semver.validRange(dependencySpecifier.versionSpecifier) &&
+            !semver.valid(dependencySpecifier.versionSpecifier)
+              ? dependencySpecifier.versionSpecifier
+              : '*';
+          packageJson.addOrUpdateDependency(name, `workspace:${workspaceRange}`, dependencyType);
+          shrinkwrapIsUpToDate = false;
+          continue;
+        } else if (dependencySpecifier.specifierType === 'workspace') {
+          // Already specified as a local project. Allow the package manager to validate this
+          continue;
+        }
+
+        // Allow the package manager to handle peer dependency resolution, since this is simply a constraint
+        // enforced by the package manager
+        if (dependencyType === DependencyType.Peer) {
+          continue;
+        }
+
+        // It is not a local dependency, validate that it is compatible
+        if (
+          shrinkwrapFile &&
+          !shrinkwrapFile.hasCompatibleWorkspaceDependency(
+            dependencySpecifier,
+            shrinkwrapFile.getWorkspaceKeyByPath(
+              this.rushConfiguration.commonTempFolder,
+              rushProject.projectFolder
+            )
+          )
+        ) {
+          shrinkwrapWarnings.push(
+            `Missing dependency "${name}" (${version}) required by "${rushProject.packageName}"`
+          );
+          shrinkwrapIsUpToDate = false;
+        }
+      }
+
+      // Save the package.json if we modified the version references and warn that the package.json was modified
+      if (packageJson.saveIfModified()) {
+        console.log(
+          colors.yellow(
+            `"${rushProject.packageName}" depends on one or more workspace packages which did not use "workspace:" ` +
+              'notation. The package.json has been modified and must be committed to source control.'
+          )
+        );
+      }
+    }
+
+    // Write the common package.json
+    InstallHelpers.generateCommonPackageJson(this.rushConfiguration);
+
+    // Save the generated workspace file. Don't update the file timestamp unless the content has changed,
+    // since "rush install" will consider this timestamp
+    workspaceFile.save(workspaceFile.workspaceFilename, { onlyIfChanged: true });
+
+    stopwatch.stop();
+    console.log(`Finished creating workspace (${stopwatch.toString()})`);
+
+    return { shrinkwrapIsUpToDate, shrinkwrapWarnings };
+  }
+
+  protected canSkipInstall(lastModifiedDate: Date): boolean {
```

Comment:
Unfortunately pnpm does not perform any checks regarding this. It will re-install from the store every time the command is run. So in this case, Rush has a bit to add in the way of usability